src/tqec/compile/compile.py

```diff
@@ -0,0 +1,307 @@
+) -> CompiledGraph:
+    """Compile a block graph.
+
+    Args:
+        block_graph: The block graph to compile.
+        custom_spec_rules: Custom specification rules for the cube specs. This is a dict
+            mapping the cube specs to the corresponding spec rules. If not provided, the
+            default spec rules will be used. Spec rules determine how to compile a cube
+            spec into a compiled block, i.e which template to use and the specific plaquettes
+            to use in the template.
+        custom_substitute_rules: Custom substitution rules for the compiled blocks. This is
+            a dict mapping the substitution keys to the corresponding substitution rules. If
+            not provided, the default substitution rules will be used. Substitution rules
+            determine how to substitute plaquettes in the two compiled blocks connected by a pipe.
+
+    Returns:
+        A `CompiledGraph` object that can be used to generate a cirq/stim circuit and scale easily.
+    """
+    if block_graph.num_open_ports != 0:
+        raise TQECException(
+            "Can not compile a block graph with open ports into circuits."
+        )
+    spec_rules = DEFAULT_SPEC_RULES | (custom_spec_rules or {})
+    substitute_rules = DEFAULT_SUBSTITUTION_RULES | (custom_substitute_rules or {})
+    cube_specs = {
+        cube: CubeSpec.from_cube(cube, block_graph) for cube in block_graph.cubes
+    }
+
+    # 1. Get the base compiled blocks before applying the substitution rules.
+    blocks: dict[Position3D, CompiledBlock] = {}
+    for cube in block_graph.cubes:
+        spec = cube_specs[cube]
+        spec_rule = spec_rules[spec]
+        blocks[cube.position] = spec_rule(spec)
+
+    # 2. Apply the substitution rules to the compiled blocks inplace.
+    pipes = block_graph.pipes
+    time_pipes = [pipe for pipe in pipes if pipe.pipe_type.direction == Direction3D.Z]
+    space_pipes = [pipe for pipe in pipes if pipe.pipe_type.direction != Direction3D.Z]
+    # Note that the order of the pipes to apply the substitution rules is important.
+    # To keep the time-direction substitution rules from removing the extra resets
+    # added by the space-direction substitution rules, we first apply the time-direction
+    # substitution rules.
+    for pipe in time_pipes + space_pipes:
+        pos1, pos2 = pipe.u.position, pipe.v.position
+        key = SubstitutionKey(cube_specs[pipe.u], cube_specs[pipe.v], pipe.pipe_type)
+        rule = substitute_rules[key]
+        blocks[pos1], blocks[pos2] = rule(key, blocks[pos1], blocks[pos2])
+
+    # 3. Collect by time and tile the blocks.
+    min_z = min(pos.z for pos in blocks.keys())
+    max_z = max(pos.z for pos in blocks.keys())
+    tiles_in_time: dict[int, TiledBlocks] = {
+        z: TiledBlocks(
+            {pos.as_2d(): block for pos, block in blocks.items() if pos.z == z}
+        )
+        for z in range(min_z, max_z + 1)
+    }
```

Comment:
Couldn't `tiles_in_time` (and so also `CompiledGraph.tiles_by_time`) be a `list[TiledBlocks]` instead? There should be no "holes" in time (else, its another computation starting from scratch, which should be in its own file) and I do not see any issue in fixing the start of time to `0`, even if the computation described in the sketchup file seems to start above/below the origin of the Z axis.

Reply:
Yes you are right, I'll make the change.